pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -518,40 +523,56 @@ private void commitSegmentMetadataInternal(String realtimeTableName,
-            .collect(Collectors.toSet());
-    int numPartitionGroups = newPartitionGroupMetadataList.size();
-
+    // Step-2
+    long startTimeNs2 = System.nanoTime();
     String newConsumingSegmentName = null;
-    if (!isTablePaused(idealState) && newPartitionGroupSet.contains(committingSegmentPartitionGroupId)) {
-      // Only if committingSegment's partitionGroup is present in the newPartitionGroupMetadataList, we create new
-      // segment metadata
-      String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
-      long newSegmentCreationTimeMs = getCurrentTimeMs();
-      LLCSegmentName newLLCSegment = new LLCSegmentName(rawTableName, committingSegmentPartitionGroupId,
-          committingLLCSegment.getSequenceNumber() + 1, newSegmentCreationTimeMs);
-      createNewSegmentZKMetadata(tableConfig, streamConfig, newLLCSegment, newSegmentCreationTimeMs,
-          committingSegmentDescriptor, committingSegmentZKMetadata, instancePartitions, numPartitionGroups, numReplicas,
-          newPartitionGroupMetadataList);
-      newConsumingSegmentName = newLLCSegment.getSegmentName();
+    if (!isTablePaused(idealState)) {
+      int numPartitionGroups;
+      boolean shouldCreateNewConsumingSegment;
+      StreamConfig streamConfig =
+          new StreamConfig(tableConfig.getTableName(), IngestionConfigUtils.getStreamConfigMap(tableConfig));
+      try {
+        numPartitionGroups = getNumPartitionGroups(streamConfig);
+        shouldCreateNewConsumingSegment = numPartitionGroups > committingSegmentPartitionGroupId;
+      } catch (Exception e) {
+        LOGGER.info("Failed to read partition count from stream metadata provider for table: {}, exception: {}. "
+                + "Reading all partition group metadata to determine partition count and partition group status.",
+            realtimeTableName, e.toString());
+        // TODO: Find a better way to determine partition count and if the committing partition group is fully consumed.
+        //       We don't need to read partition group metadata for other partition groups.
+        List<PartitionGroupConsumptionStatus> currentPartitionGroupConsumptionStatusList =
+            getPartitionGroupConsumptionStatusList(idealState, streamConfig);
+        List<PartitionGroupMetadata> newPartitionGroupMetadataList =
+            getNewPartitionGroupMetadataList(streamConfig, currentPartitionGroupConsumptionStatusList);
+        numPartitionGroups = newPartitionGroupMetadataList.size();
+        shouldCreateNewConsumingSegment = false;
+        for (PartitionGroupMetadata newPartitionGroupMetadata : newPartitionGroupMetadataList) {
+          if (newPartitionGroupMetadata.getPartitionGroupId() == committingSegmentPartitionGroupId) {
+            shouldCreateNewConsumingSegment = true;
+            break;
+          }
+        }
+      }
+      if (shouldCreateNewConsumingSegment) {
```

Comment:
Just double checking, nothing here is expecting partitionIds to be monotnically increasing right?

Reply:
I don't fully follow. What is your concern here?

Reply:
the default implementation in the [StreamMetadataProvider](https://github.com/apache/pinot/blob/a986dd15368c0554902c6cca571d0b60a084b547/pinot-spi/src/main/java/org/apache/pinot/spi/stream/StreamMetadataProvider.java#L82) assumes partitionIds are 0..partitionCount. That's not the case for our partitions. I just want to make sure that's not a problem here.

Reply:
That is indeed a problem. Can you elaborate more on the partitionIds in your stream metadata provider?

Reply:
since we consume from multiple kafka clusters, our partitionIds are a custom encoding with half the bits representing the kafka cluster and the other half representing the partition number.

Our "partitioncount" returns the total partitions across all clusters. So we do expect `fetchPartitionCount` to increase. It's just that the IDs themselves aren't ordered.

Reply:
@jadami10 Added `Set<Integer> fetchPartitionIds(long timeoutMillis)` into `StreamMetadataProvider` with a default implementation. Let me know if it suits your need